_posts/2021-01-27-create-analyze-heapdump.md

```diff
@@ -0,0 +1,241 @@
+
+![PieChart](/assets/img/posts/heapdump/piechart.png)
+
+For our application, this information in the overview means if we could dispose of a particular instance of `java.lang.Thread` we will save 1.7 GB, and almost all of the memory used in this application. 
+
+#### Histogram View
+While that might look promising, java.lang.Thread is unlikely to be the real problem here. To get a better insight into what objects currently exist, we will use the Histogram view:
+
+![histogram](/assets/img/posts/heapdump/histogram.png)
+
+We have filtered the histogram with a regular expression "io.pratik.* " to show only the classes that match the pattern. With this view, we can see the number of live objects: for example, 243 `BrandedProduct` objects, and 309 `Price` Objects are alive in the system. We can also see the amount of memory each object is using. 
+
+There are two calculations, Shallow Heap and Retained Heap.  A shallow heap is the amount of memory consumed by one object. An Object requires 32 (or 64 bits, depending on the architecture) for each reference. Primitives such as integers and longs require 4 or 8 bytes, etc… While this can be interesting, the more useful metric is the Retained Heap.
+
+#### Retained Heap Size
+The retained heap size is computed by adding the size of all the objects in the retained set. A retained set of X is the set of objects which would be removed by the GC when X is collected.
+
+The retained heap can be calculated in two different ways, using the quick approximation or the precise retained size:
+
+![retainedheap](/assets/img/posts/heapdump/retainedheap.png)
+
+By calculating the Retained Heap we can now see that io.pratik.ProductGroup is holding the majority of the memory, even though it is only 32 bytes(shallow heap size) by itself. By finding a way to free up this object, we can certainly get our memory problem under control.
+
+#### Dominator Tree
+The dominator tree is used to identify the retained heap. It is produced by the complex object graph generated at runtime and helps to identify the largest memory graphs. An Object X is said to dominate an Object Y if every path from the Root to Y must pass through X. 
+
+Looking at the dominator tree for our example, we can see which objects are retained in the memory.
+![dominatortree](/assets/img/posts/heapdump/dominatortree.png)
+We can see that the `ProductGroup` object holds the memory instead of the `Thread` object. We can probably fix the memory problem by releasing objects contained in this object. 
+
+
+#### Leak Suspects Report
+We can also generate a "Leak Suspects Report" to find a suspected big object or set of objects. This report presents the findings on an HTML page and is also saved in a zip file next to the heap dump file. 
+
+Due to its smaller size, it is preferable to share the "Leak Suspects Report" report with teams specialized in performing analysis tasks instead of the raw heap dump file.
```

Comment:
But the leak suspects report only shows the `Thread` as a suspect in this case and not the `ProductGroup` which is the real problem. If we only share the report, the analyzing team will not have all the information to find the actual cause, will it?